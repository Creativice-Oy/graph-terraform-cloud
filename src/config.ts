import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationProviderAuthenticationError,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { TerraformCloudClient } from './tfe/client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  apiKey: {
    type: 'string',
  },
  organizationOwner: {
    type: 'string',
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  apiKey: string;
  organizationOwner: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { instance } = context;
  const { config } = instance;

  if (!config.apiKey) {
    throw new IntegrationValidationError('Config requires all of {apiKey}');
  }

  if (!config.organizationOwner) {
    throw new IntegrationValidationError(
      "Please indicate if you're an organization owner.",
    );
  }

  const client = new TerraformCloudClient({ apiKey: config.apiKey });

  try {
    await client.organizations.list();
  } catch (err) {
    throw new IntegrationProviderAuthenticationError(err.message);
  }
}
