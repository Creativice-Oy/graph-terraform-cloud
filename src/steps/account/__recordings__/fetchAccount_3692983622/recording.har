{
  "log": {
    "_recordingName": "fetchAccount",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "020c85ae52b49216a932f1ac0762939b",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/vnd.api+json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.terraform.io"
            }
          ],
          "headersSize": 373,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.terraform.io/api/v2/account/details"
        },
        "response": {
          "bodySize": 774,
          "content": {
            "mimeType": "application/vnd.api+json; charset=utf-8",
            "size": 774,
            "text": "{\"data\":{\"id\":\"user-2yZJWZGNJmNsidGK\",\"type\":\"users\",\"attributes\":{\"username\":\"pgilaga\",\"is-service-account\":false,\"avatar-url\":\"https://www.gravatar.com/avatar/6bf8d79d3b1d6d34a9d69f62cd09d20a?s=100\\u0026d=mm\",\"password\":null,\"enterprise-support\":false,\"is-site-admin\":false,\"is-sso-login\":false,\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"pgilaga@gmail.com\",\"unconfirmed-email\":null,\"has-git-hub-app-token\":false,\"is-confirmed\":true,\"is-sudo\":false,\"permissions\":{\"can-create-organizations\":true,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":true}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-2yZJWZGNJmNsidGK/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-2yZJWZGNJmNsidGK\"}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 31 Dec 2021 04:02:34 GMT"
            },
            {
              "name": "content-type",
              "value": "application/vnd.api+json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "774"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "etag",
              "value": "W/\"83c358495801c34429348005e9d40296\""
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "tfp-api-version",
              "value": "2.5"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Origin"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-download-options",
              "value": "noopen"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-permitted-cross-domain-policies",
              "value": "none"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "30"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "29"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "0.03"
            },
            {
              "name": "x-request-id",
              "value": "a57b36f3-618e-79d8-5d33-acfbbd7a87b7"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 697,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-12-31T04:02:34.040Z",
        "time": 991,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 991
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
