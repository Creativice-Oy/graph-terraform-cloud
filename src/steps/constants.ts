import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const ACCOUNT_ENTITY_KEY = 'ACCOUNT_ENTITY';
export const ACCOUNT_DATA_KEY = 'ACCOUNT_DATA';
export const MAC_OS_CONFIGURATION_DETAILS_BY_ID_KEY =
  'MAC_OS_CONFIGURATION_DETAILS_BY_ID';

export enum IntegrationSteps {
  ORGANIZATIONS = 'fetch-organizations',
  ORGANIZATION_MEMBERS = 'fetch-organization-members',
  ORGANIZATION_OAUTH_TOKENS = 'fetch-organization-oauth-tokens',
}

export const Entities: Record<'ORGANIZATION' | 'USER', StepEntityMetadata> = {
  ORGANIZATION: {
    _type: 'tfe_organization',
    _class: ['Account', 'Organization'],
    resourceName: 'Organization',
  },
  USER: {
    _type: 'tfe_user',
    _class: ['User'],
    resourceName: 'User',
  },
};

export const Relationships: Record<
  'ORGANIZATION_HAS_USER',
  StepRelationshipMetadata
> = {
  ORGANIZATION_HAS_USER: {
    _type: 'tfe_organization_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ORGANIZATION._type,
    targetType: Entities.USER._type,
  },
};
