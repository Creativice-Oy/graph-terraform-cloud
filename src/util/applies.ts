import fetch from 'node-fetch';

/**
 * See: https://www.terraform.io/docs/internals/machine-readable-ui.html#operation-messages
 */
export type ParsedTerraformApplyLogActionType =
  | 'apply_start'
  | 'apply_progress'
  | 'apply_complete'
  | 'apply_errored'
  | 'provision_start'
  | 'provision_progress'
  | 'provision_complete'
  | 'provision_errored'
  | 'refresh_start'
  | 'refresh_complete';

export type ParsedTerraformApplyLogHook = {
  resource: {
    addr: string;
    module: string;
    resource: string;
    implied_provider: string;
    resource_type: string;
    resource_name: string;
    resource_key: string | null;
  };
  action: string;
  id_key: string;
  id_value: string;
  elapsed_seconds: number;
};

export type ParsedTerraformApplyLog = {
  '@level': string;
  '@message': string;
  '@module': string;
  '@timestamp': string;
  type: string;
  hook: ParsedTerraformApplyLogHook;
};

export async function downloadAndParseRunApply(url: string) {
  const response = await fetch(url);
  return response.text();
}

/**
 * When the `Structured Run Output` workspace option is enabled, Terraform Cloud
 * stores raw logs of the Terraform apply.
 */
export function parseRunApply(text: string): ParsedTerraformApplyLog[] {
  const splitApply = text.split('\n');
  const collected: ParsedTerraformApplyLog[] = [];

  for (const row of splitApply) {
    try {
      const parsed = JSON.parse(row) as ParsedTerraformApplyLog;

      // Validate that one of the apply log properties are available before
      // considering it valid
      if (parsed['@message']) {
        collected.push(parsed);
      }
    } catch (err) {
      // This apply log is likely using non-structured run output and therefore
      // cannot be parsed
    }
  }
  return collected;
}

export async function collectApplyCompleteOperations(url: string) {
  const text = await downloadAndParseRunApply(url);
  const parsed = parseRunApply(text);
  return parsed.filter((v) => v.type === 'apply_complete');
}
