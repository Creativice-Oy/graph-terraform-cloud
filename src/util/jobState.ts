import { JobState } from '@jupiterone/integration-sdk-core';

const ORGANIZATION_IDS_CACHE_KEY = 'tfe_organization_ids';

export interface CachedOrganizationData {
  organizationName: string;
  organizationExternalId: string;
}

export async function cacheOrganizationData(
  jobState: JobState,
  organizationData: CachedOrganizationData[],
) {
  await jobState.setData(ORGANIZATION_IDS_CACHE_KEY, organizationData);
}

export async function getOrganizationsFromCache(
  jobState: JobState,
): Promise<CachedOrganizationData[]> {
  const data = await jobState.getData<CachedOrganizationData[]>(
    ORGANIZATION_IDS_CACHE_KEY,
  );

  return data || [];
}

export async function forEachOrganization(
  jobState: JobState,
  callback: (organization: CachedOrganizationData) => Promise<void>,
) {
  for (const result of await getOrganizationsFromCache(jobState)) {
    await callback(result);
  }
}
