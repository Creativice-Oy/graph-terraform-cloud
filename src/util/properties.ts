export function prefixObjProperties(prefix: string, obj: Record<string, any>) {
  const collected: Record<string, any> = {};

  for (const key in obj) {
    const val = obj[key];
    collected[`${prefix}.${key}`] = val;
  }

  return collected;
}
