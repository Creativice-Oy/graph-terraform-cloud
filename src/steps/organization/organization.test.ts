import { fetchOrganizationMembers, fetchOrganizations } from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { RelationshipClass } from '@jupiterone/data-model';
import { cacheOrganizationData } from '../../util/jobState';
import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';

describe('#fetchOrganizations', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchOrganizations',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchOrganizations],
      entitySchemaMatchers: [
        {
          _type: Entities.ORGANIZATION._type,
          matcher: {
            _class: ['Account', 'Organization'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'tfe_organization' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                email: { type: 'string' },
                collaboratorAuthPolicy: { type: 'string' },
                ownersTeamSamlRoleId: { type: ['string', 'null'] },
                'permissions.canAccessViaTeams': { type: 'boolean' },
                'permissions.canCreateModule': { type: 'boolean' },
                'permissions.canCreateProvider': { type: 'boolean' },
                'permissions.canCreateTeam': { type: 'boolean' },
                'permissions.canCreateWorkspace': { type: 'boolean' },
                'permissions.canDestroy': { type: 'boolean' },
                'permissions.canManagePublicModules': { type: 'boolean' },
                'permissions.canManagePublicProviders': { type: 'boolean' },
                'permissions.canManageSso': { type: 'boolean' },
                'permissions.canManageSubscription': { type: 'boolean' },
                'permissions.canManageTags': { type: 'boolean' },
                'permissions.canManageUsers': { type: 'boolean' },
                'permissions.canStartTrial': { type: 'boolean' },
                'permissions.canTraverse': { type: 'boolean' },
                'permissions.canUpdate': { type: 'boolean' },
                'permissions.canUpdateAgentPools': { type: 'boolean' },
                'permissions.canUpdateApiToken': { type: 'boolean' },
                'permissions.canUpdateOauth': { type: 'boolean' },
                'permissions.canUpdateSentinel': { type: 'boolean' },
                'permissions.canUpdateSshKeys': { type: 'boolean' },
                planExpired: { type: 'boolean' },
                planExpiresAt: { type: 'string' },
                planIsEnterprise: { type: 'boolean' },
                planIsTrial: { type: 'boolean' },
                samlEnabled: { type: 'boolean' },
                sessionRemember: { type: ['number', 'null'] },
                sessionTimeout: { type: ['number', 'null'] },
                costEstimationEnabled: { type: 'boolean' },
                fairRunQueuingEnabled: { type: 'boolean' },
                twoFactorConformant: { type: 'boolean' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [],
    });
  });
});

describe('#fetchOrganizationMembers', () => {
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await cacheOrganizationData(context.jobState, [
      {
        organizationName: 'austin-test-org',
        organizationExternalId: 'org-Ljx2Ap3vszNhHWWv',
      },
      {
        organizationName: 'austin-test-org-v2',
        organizationExternalId: 'org-X2NET8eXZP36vxN7',
      },
      {
        organizationName: 'jupiterone',
        organizationExternalId: 'org-yMai1ZUhS39WMAU9',
      },
    ]);

    await createDataCollectionTest({
      context,
      recordingName: 'fetchOrganizationMembers',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchOrganizationMembers],
      entitySchemaMatchers: [
        {
          _type: Entities.USER._type,
          matcher: {
            _class: ['User'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'tfe_user' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                email: { type: 'string' },
                isServiceAccount: { type: 'boolean' },
                avatarUrl: { type: 'string' },
                mfaEnabled: { type: 'boolean' },
                mfaVerified: { type: 'boolean' },
                'permissions.canChangeEmail': { type: 'boolean' },
                'permissions.canCreateOrganizations': { type: 'boolean' },
                'permissions.canManageUserTokens': { type: 'boolean' },
                'permissions.canChangeUsername': { type: 'boolean' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.ORGANIZATION_HAS_USER._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.HAS },
                _type: { const: 'tfe_organization_has_user' },
                status: { type: 'string' },
              },
            },
          },
        },
      ],
    });
  });
});
