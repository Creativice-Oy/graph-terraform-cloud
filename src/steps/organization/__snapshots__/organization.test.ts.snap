// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchOrganizationMembers should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "user-dtoAE5LXupfvycjK",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "avatarUrl": "https://www.gravatar.com/avatar/1941eaaff412c1a04cfd4166bfd66371?s=100&d=mm",
            "email": "austin.kelleher@jupiterone.com",
            "isServiceAccount": false,
            "permissions": Object {
              "canChangeEmail": true,
              "canChangeUsername": true,
              "canCreateOrganizations": true,
              "canManageUserTokens": true,
            },
            "twoFactor": Object {
              "enabled": false,
              "verified": false,
            },
            "username": "austinkelleher-jupiterone",
          },
        },
      ],
      "_type": "tfe_user",
      "avatarUrl": "https://www.gravatar.com/avatar/1941eaaff412c1a04cfd4166bfd66371?s=100&d=mm",
      "createdOn": undefined,
      "displayName": "austinkelleher-jupiterone",
      "email": "austin.kelleher@jupiterone.com",
      "id": "user-dtoAE5LXupfvycjK",
      "isServiceAccount": false,
      "mfaEnabled": false,
      "mfaVerified": false,
      "name": "austinkelleher-jupiterone",
      "permissions.canChangeEmail": true,
      "permissions.canChangeUsername": true,
      "permissions.canCreateOrganizations": true,
      "permissions.canManageUserTokens": true,
      "username": "austinkelleher-jupiterone",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "user-HMyUEwhb3mB45HZ7",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "avatarUrl": "https://www.gravatar.com/avatar/4cdbb2e21e2d991b0dd11d9a38c2c660?s=100&d=mm",
            "email": "nick.dowmon@jupiterone.com",
            "isServiceAccount": false,
            "permissions": Object {
              "canChangeEmail": true,
              "canChangeUsername": true,
              "canCreateOrganizations": true,
              "canManageUserTokens": false,
            },
            "twoFactor": Object {
              "enabled": false,
              "verified": false,
            },
            "username": "ndowmon",
          },
        },
      ],
      "_type": "tfe_user",
      "avatarUrl": "https://www.gravatar.com/avatar/4cdbb2e21e2d991b0dd11d9a38c2c660?s=100&d=mm",
      "createdOn": undefined,
      "displayName": "ndowmon",
      "email": "nick.dowmon@jupiterone.com",
      "id": "user-HMyUEwhb3mB45HZ7",
      "isServiceAccount": false,
      "mfaEnabled": false,
      "mfaVerified": false,
      "name": "ndowmon",
      "permissions.canChangeEmail": true,
      "permissions.canChangeUsername": true,
      "permissions.canCreateOrganizations": true,
      "permissions.canManageUserTokens": false,
      "username": "ndowmon",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "user-2yZJWZGNJmNsidGK",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "avatarUrl": "https://www.gravatar.com/avatar/6bf8d79d3b1d6d34a9d69f62cd09d20a?s=100&d=mm",
            "email": "pgilaga@gmail.com",
            "isServiceAccount": false,
            "permissions": Object {
              "canChangeEmail": true,
              "canChangeUsername": true,
              "canCreateOrganizations": true,
              "canManageUserTokens": false,
            },
            "twoFactor": Object {
              "enabled": false,
              "verified": false,
            },
            "username": "pgilaga",
          },
        },
      ],
      "_type": "tfe_user",
      "avatarUrl": "https://www.gravatar.com/avatar/6bf8d79d3b1d6d34a9d69f62cd09d20a?s=100&d=mm",
      "createdOn": undefined,
      "displayName": "pgilaga",
      "email": "pgilaga@gmail.com",
      "id": "user-2yZJWZGNJmNsidGK",
      "isServiceAccount": false,
      "mfaEnabled": false,
      "mfaVerified": false,
      "name": "pgilaga",
      "permissions.canChangeEmail": true,
      "permissions.canChangeUsername": true,
      "permissions.canCreateOrganizations": true,
      "permissions.canManageUserTokens": false,
      "username": "pgilaga",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "user-qZQNfnC19eG6Ameo",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "avatarUrl": "https://www.gravatar.com/avatar/5a764b0f770a7bfe74cdc94afea04ed6?s=100&d=mm",
            "email": "stefan@creativice.com",
            "isServiceAccount": false,
            "permissions": Object {
              "canChangeEmail": true,
              "canChangeUsername": true,
              "canCreateOrganizations": true,
              "canManageUserTokens": false,
            },
            "twoFactor": Object {
              "enabled": false,
              "verified": false,
            },
            "username": "stefan-creativice",
          },
        },
      ],
      "_type": "tfe_user",
      "avatarUrl": "https://www.gravatar.com/avatar/5a764b0f770a7bfe74cdc94afea04ed6?s=100&d=mm",
      "createdOn": undefined,
      "displayName": "stefan-creativice",
      "email": "stefan@creativice.com",
      "id": "user-qZQNfnC19eG6Ameo",
      "isServiceAccount": false,
      "mfaEnabled": false,
      "mfaVerified": false,
      "name": "stefan-creativice",
      "permissions.canChangeEmail": true,
      "permissions.canChangeUsername": true,
      "permissions.canCreateOrganizations": true,
      "permissions.canManageUserTokens": false,
      "username": "stefan-creativice",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-Ljx2Ap3vszNhHWWv",
      "_key": "org-Ljx2Ap3vszNhHWWv|has|user-dtoAE5LXupfvycjK",
      "_toEntityKey": "user-dtoAE5LXupfvycjK",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-X2NET8eXZP36vxN7",
      "_key": "org-X2NET8eXZP36vxN7|has|user-dtoAE5LXupfvycjK",
      "_toEntityKey": "user-dtoAE5LXupfvycjK",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-yMai1ZUhS39WMAU9",
      "_key": "org-yMai1ZUhS39WMAU9|has|user-dtoAE5LXupfvycjK",
      "_toEntityKey": "user-dtoAE5LXupfvycjK",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-yMai1ZUhS39WMAU9",
      "_key": "org-yMai1ZUhS39WMAU9|has|user-HMyUEwhb3mB45HZ7",
      "_toEntityKey": "user-HMyUEwhb3mB45HZ7",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-yMai1ZUhS39WMAU9",
      "_key": "org-yMai1ZUhS39WMAU9|has|user-2yZJWZGNJmNsidGK",
      "_toEntityKey": "user-2yZJWZGNJmNsidGK",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "org-yMai1ZUhS39WMAU9",
      "_key": "org-yMai1ZUhS39WMAU9|has|user-qZQNfnC19eG6Ameo",
      "_toEntityKey": "user-qZQNfnC19eG6Ameo",
      "_type": "tfe_organization_has_user",
      "displayName": "HAS",
      "status": "active",
    },
  ],
  "encounteredTypes": Array [
    "tfe_user",
    "tfe_organization_has_user",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 6,
}
`;

exports[`#fetchOrganizations should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
        "Organization",
      ],
      "_key": "org-Ljx2Ap3vszNhHWWv",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "collaboratorAuthPolicy": "password",
            "costEstimationEnabled": false,
            "createdAt": "2021-03-28T12:46:34.199Z",
            "email": "austin.kelleher@jupiterone.com",
            "externalId": "org-Ljx2Ap3vszNhHWWv",
            "fairRunQueuingEnabled": true,
            "name": "austin-test-org",
            "ownersTeamSamlRoleId": null,
            "permissions": Object {
              "canAccessViaTeams": true,
              "canCreateModule": true,
              "canCreateProvider": false,
              "canCreateTeam": false,
              "canCreateWorkspace": true,
              "canDestroy": true,
              "canManagePublicModules": true,
              "canManagePublicProviders": false,
              "canManageSso": false,
              "canManageSubscription": true,
              "canManageTags": true,
              "canManageUsers": true,
              "canStartTrial": true,
              "canTraverse": true,
              "canUpdate": true,
              "canUpdateAgentPools": false,
              "canUpdateApiToken": true,
              "canUpdateOauth": true,
              "canUpdateSentinel": false,
              "canUpdateSshKeys": true,
            },
            "planExpired": false,
            "planExpiresAt": null,
            "planIsEnterprise": false,
            "planIsTrial": false,
            "samlEnabled": false,
            "sessionRemember": null,
            "sessionTimeout": null,
            "twoFactorConformant": false,
          },
        },
      ],
      "_type": "tfe_organization",
      "collaboratorAuthPolicy": "password",
      "costEstimationEnabled": false,
      "createdOn": 1616935594199,
      "displayName": "austin-test-org",
      "email": "austin.kelleher@jupiterone.com",
      "fairRunQueuingEnabled": true,
      "id": "org-Ljx2Ap3vszNhHWWv",
      "mfaEnabled": false,
      "name": "austin-test-org",
      "ownersTeamSamlRoleId": null,
      "permissions.canAccessViaTeams": true,
      "permissions.canCreateModule": true,
      "permissions.canCreateProvider": false,
      "permissions.canCreateTeam": false,
      "permissions.canCreateWorkspace": true,
      "permissions.canDestroy": true,
      "permissions.canManagePublicModules": true,
      "permissions.canManagePublicProviders": false,
      "permissions.canManageSso": false,
      "permissions.canManageSubscription": true,
      "permissions.canManageTags": true,
      "permissions.canManageUsers": true,
      "permissions.canStartTrial": true,
      "permissions.canTraverse": true,
      "permissions.canUpdate": true,
      "permissions.canUpdateAgentPools": false,
      "permissions.canUpdateApiToken": true,
      "permissions.canUpdateOauth": true,
      "permissions.canUpdateSentinel": false,
      "permissions.canUpdateSshKeys": true,
      "planExpired": false,
      "planExpiresAt": undefined,
      "planIsEnterprise": false,
      "planIsTrial": false,
      "samlEnabled": false,
      "sessionRemember": null,
      "sessionTimeout": null,
      "webLink": "https://app.terraform.io/app/austin-test-org/workspaces",
    },
    Object {
      "_class": Array [
        "Account",
        "Organization",
      ],
      "_key": "org-X2NET8eXZP36vxN7",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "collaboratorAuthPolicy": "password",
            "costEstimationEnabled": false,
            "createdAt": "2021-07-18T12:37:48.821Z",
            "email": "austin.kelleher@jupiterone.com",
            "externalId": "org-X2NET8eXZP36vxN7",
            "fairRunQueuingEnabled": true,
            "name": "austin-test-org-v2",
            "ownersTeamSamlRoleId": null,
            "permissions": Object {
              "canAccessViaTeams": true,
              "canCreateModule": true,
              "canCreateProvider": false,
              "canCreateTeam": false,
              "canCreateWorkspace": true,
              "canDestroy": true,
              "canManagePublicModules": true,
              "canManagePublicProviders": false,
              "canManageSso": false,
              "canManageSubscription": true,
              "canManageTags": true,
              "canManageUsers": true,
              "canStartTrial": true,
              "canTraverse": true,
              "canUpdate": true,
              "canUpdateAgentPools": false,
              "canUpdateApiToken": true,
              "canUpdateOauth": true,
              "canUpdateSentinel": false,
              "canUpdateSshKeys": true,
            },
            "planExpired": false,
            "planExpiresAt": null,
            "planIsEnterprise": false,
            "planIsTrial": false,
            "samlEnabled": false,
            "sessionRemember": null,
            "sessionTimeout": null,
            "twoFactorConformant": false,
          },
        },
      ],
      "_type": "tfe_organization",
      "collaboratorAuthPolicy": "password",
      "costEstimationEnabled": false,
      "createdOn": 1626611868821,
      "displayName": "austin-test-org-v2",
      "email": "austin.kelleher@jupiterone.com",
      "fairRunQueuingEnabled": true,
      "id": "org-X2NET8eXZP36vxN7",
      "mfaEnabled": false,
      "name": "austin-test-org-v2",
      "ownersTeamSamlRoleId": null,
      "permissions.canAccessViaTeams": true,
      "permissions.canCreateModule": true,
      "permissions.canCreateProvider": false,
      "permissions.canCreateTeam": false,
      "permissions.canCreateWorkspace": true,
      "permissions.canDestroy": true,
      "permissions.canManagePublicModules": true,
      "permissions.canManagePublicProviders": false,
      "permissions.canManageSso": false,
      "permissions.canManageSubscription": true,
      "permissions.canManageTags": true,
      "permissions.canManageUsers": true,
      "permissions.canStartTrial": true,
      "permissions.canTraverse": true,
      "permissions.canUpdate": true,
      "permissions.canUpdateAgentPools": false,
      "permissions.canUpdateApiToken": true,
      "permissions.canUpdateOauth": true,
      "permissions.canUpdateSentinel": false,
      "permissions.canUpdateSshKeys": true,
      "planExpired": false,
      "planExpiresAt": undefined,
      "planIsEnterprise": false,
      "planIsTrial": false,
      "samlEnabled": false,
      "sessionRemember": null,
      "sessionTimeout": null,
      "webLink": "https://app.terraform.io/app/austin-test-org-v2/workspaces",
    },
    Object {
      "_class": Array [
        "Account",
        "Organization",
      ],
      "_key": "org-yMai1ZUhS39WMAU9",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "collaboratorAuthPolicy": "password",
            "costEstimationEnabled": false,
            "createdAt": "2021-02-17T21:24:28.929Z",
            "email": "austin.kelleher@jupiterone.com",
            "externalId": "org-yMai1ZUhS39WMAU9",
            "fairRunQueuingEnabled": true,
            "name": "jupiterone",
            "ownersTeamSamlRoleId": null,
            "permissions": Object {
              "canAccessViaTeams": true,
              "canCreateModule": true,
              "canCreateProvider": false,
              "canCreateTeam": false,
              "canCreateWorkspace": true,
              "canDestroy": true,
              "canManagePublicModules": true,
              "canManagePublicProviders": false,
              "canManageSso": false,
              "canManageSubscription": true,
              "canManageTags": true,
              "canManageUsers": true,
              "canStartTrial": true,
              "canTraverse": true,
              "canUpdate": true,
              "canUpdateAgentPools": false,
              "canUpdateApiToken": true,
              "canUpdateOauth": true,
              "canUpdateSentinel": false,
              "canUpdateSshKeys": true,
            },
            "planExpired": false,
            "planExpiresAt": null,
            "planIsEnterprise": false,
            "planIsTrial": false,
            "samlEnabled": false,
            "sessionRemember": null,
            "sessionTimeout": null,
            "twoFactorConformant": false,
          },
        },
      ],
      "_type": "tfe_organization",
      "collaboratorAuthPolicy": "password",
      "costEstimationEnabled": false,
      "createdOn": 1613597068929,
      "displayName": "jupiterone",
      "email": "austin.kelleher@jupiterone.com",
      "fairRunQueuingEnabled": true,
      "id": "org-yMai1ZUhS39WMAU9",
      "mfaEnabled": false,
      "name": "jupiterone",
      "ownersTeamSamlRoleId": null,
      "permissions.canAccessViaTeams": true,
      "permissions.canCreateModule": true,
      "permissions.canCreateProvider": false,
      "permissions.canCreateTeam": false,
      "permissions.canCreateWorkspace": true,
      "permissions.canDestroy": true,
      "permissions.canManagePublicModules": true,
      "permissions.canManagePublicProviders": false,
      "permissions.canManageSso": false,
      "permissions.canManageSubscription": true,
      "permissions.canManageTags": true,
      "permissions.canManageUsers": true,
      "permissions.canStartTrial": true,
      "permissions.canTraverse": true,
      "permissions.canUpdate": true,
      "permissions.canUpdateAgentPools": false,
      "permissions.canUpdateApiToken": true,
      "permissions.canUpdateOauth": true,
      "permissions.canUpdateSentinel": false,
      "permissions.canUpdateSshKeys": true,
      "planExpired": false,
      "planExpiresAt": undefined,
      "planIsEnterprise": false,
      "planIsTrial": false,
      "samlEnabled": false,
      "sessionRemember": null,
      "sessionTimeout": null,
      "webLink": "https://app.terraform.io/app/jupiterone/workspaces",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "tfe_organization",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 0,
}
`;
